src/source/nats.rs

```diff
@@ -0,0 +1,178 @@
+            uid,
+            scheme: "tremor-nats".to_string(),
+            host: "not-connected".to_string(),
+            port: None,
+            path: vec![],
+        };
+        Self {
+            onramp_id,
+            config,
+            subscription: None,
+            connection: None,
+            origin_uri,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for Int {
+    async fn pull_event(&mut self, _id: u64) -> Result<SourceReply> {
+        if let Some(sub) = &self.subscription {
+            if let Some(msg) = sub.next().await {
+                let mut origin_uri = self.origin_uri.clone();
+                origin_uri.path = vec![msg.subject];
+                let data = msg.data;
+                let mut nats_meta_data = Value::object_with_capacity(1);
+                let mut msg_reply = None;
+                if let Some(reply) = msg.reply {
+                    msg_reply = Some(reply);
+                }
+                let mut msg_headers = None;
+                if let Some(headers) = msg.headers {
+                    let mut key_val: Value = Value::object_with_capacity(headers.len());
+                    for (key, val) in headers.iter() {
+                        let key = String::from(key);
+                        let val: Vec<String> = val.into_iter().map(|x| String::from(x)).collect();
+                        key_val.insert(key, val)?;
+                    }
+                    msg_headers = Some(key_val);
+                }
+                let mut meta_data = Value::object_with_capacity(2);
+                if let Some(msg_reply) = msg_reply {
+                    meta_data.insert("reply", msg_reply)?;
+                }
+                if let Some(msg_headers) = msg_headers {
+                    meta_data.insert("headers", msg_headers)?;
+                }
+                nats_meta_data.insert("nats", meta_data)?;
+                Ok(SourceReply::Data {
+                    origin_uri,
+                    data,
+                    meta: Some(nats_meta_data),
+                    codec_override: None,
+                    stream: 0
+                })
+            } else {
+                Ok(SourceReply::Empty(0))
+            }
+        } else {
+           Ok(SourceReply::StateChange(SourceState::Disconnected))
+        }
+    }
+
+    fn id(&self) -> &TremorURL {
+        &self.onramp_id
+    }
+
+    async fn init(&mut self) -> Result<SourceState> {
+        let nc = self.config.connection().await?;
+        let sub = if let Some(queue) = &self.config.queue {
+            nc.queue_subscribe(self.config.subject.as_str(), queue.as_str()).await?
+        } else {
+            nc.subscribe(self.config.subject.as_str()).await?
+        };
+        self.connection = Some(nc);
+        self.subscription = Some(sub);
+        Ok(SourceState::Connected)
+    }
+
+    async fn terminate(&mut self) {
+        if let Some(connection) = &self.connection {
```

Comment:
should we drain the subscription here?

Reply:
I was in two minds about this, but decided to not do it, after going through the implementation of the [`close()`](https://github.com/nats-io/nats.rs/blob/master/src/client.rs#L222-L255) of the `nats.rs` crate, which does drains all subscriptions.